### PR TITLE
Sane add plugin

### PR DESCRIPTION
**Pull request by mkoistinen**
For posterity, this is the latest of a series of PRs for this feature: #3804 and #3516. All credit for this work goes to @ojii! Thank you!

### CODE REVIEW

File: cms/static/cms/js/modules/cms.plugins.js [hunk 1]
```diff
@@ -622,8 +578,8 @@
                         }
                     }
 
-                    // bind arrow up and shift+tab keys
-                    if (e.keyCode === 38 || (e.keyCode === 9 && e.shiftKey)) {
+                    // bind arrow up keys
+                    if (e.keyCode === 38) {
```
Comment (vxsx) [inline cms/static/cms/js/modules/cms.plugins.js]: why remove the shift+tab combo?

File: cms/static/cms/js/modules/cms.plugins.js [hunk 2]
```diff
@@ -756,37 +713,11 @@
                     if (el.hasClass('cms-dragitem-expanded')) {
                         settings.states.splice($.inArray(id, settings.states), 1);
                         el.removeClass('cms-dragitem-expanded').parent().find('> .cms-draggables').hide();
-                        if ($(document).data('expandmode')) {
```
Comment (vxsx) [inline cms/static/cms/js/modules/cms.plugins.js]: why remove shift+click on a plugin?

File: cms/static/cms/js/modules/cms.plugins.js [hunk 3]
```diff
@@ -187,8 +179,8 @@
                     that.editPlugin(
                         that.options.urls.edit_plugin,
                         that.options.plugin_name,
-                        that.options.plugin_breadcrumb
-                    );
+                        that.options.plugin_breadcrumb)
+                    ;
```
Comment (vxsx) [inline cms/static/cms/js/modules/cms.plugins.js]: this is a weird change :)

**Comment by evildmp** [comment]
Where can we find a Selenium/Travis magician?

**Comment by czpython** [comment]
In order to avoid the endless pull requests, I've created https://github.com/divio/django-cms/pull/5073 to track the progress of merging develop into this branch. Once that's completed, then I'll merge the PR into this branch.

File: cms/test_utils/project/nonroot_urls.py [hunk 4]
```diff
@@ -21,7 +21,11 @@
 
 urlpatterns += i18n_patterns(
     url(r'^admin/', include(admin.site.urls)),
-    url(r'^content/', include('cms.urls')),
+    url(r'^content/', include(
+        'cms.urls',
+        app_name=get_cms_setting('APP_NAME'),
```
Comment (czpython) [inline cms/test_utils/project/nonroot_urls.py]: Hey @ojii, is this a valid setting? can't find it on the cms and currently fails the tests

File: cms/plugin_base.py [hunk 5]
```diff
@@ -214,6 +317,16 @@ def save_model(self, request, obj, form, change):
...
                 setattr(obj, field.name, value)
+        # When adding an object, it won't have a position
+        if not obj.position:
+            if obj.parent_id:
+                obj.position = CMSPlugin.objects.filter(
+                    parent_id=obj.parent_id
+                ).count()
+            else:
+                obj.position = CMSPlugin.objects.filter(
+                    placeholder_id=obj.placeholder_id
```
Comment (czpython) [inline cms/plugin_base.py]: not considering language here. will open separate pr with a fix though

File: cms/test_utils/project/nonroot_urls.py (hunk 4 above)
Comment (yakky) [inline cms/test_utils/project/nonroot_urls.py]: This has been removed in #4917

Comment (czpython) [inline cms/test_utils/project/nonroot_urls.py]: Thanks @yakky, I will then push a fix for this in a separate branch along with other fixes.

**Comment by czpython** [comment]
Status update:

Tests are passing but ckeditor is still broken.
So to avoid endless rebasing, we'll focus on fixing ckeditor and once that's done then we'll rebase this once again with develop and then merge it.

/cc @divio/django-cms-core

**Comment by yakky** [comment]
@czpython are you aware that text plugin needs this https://github.com/divio/djangocms-text-ckeditor/pull/179 to be reapplied?

**Comment by czpython** [comment]
@yakky yup. sadly that doesn't fix the problem entirely because it doesn't cover child plugins.

**Comment by mkoistinen** [comment]
OK to merge since its already been reviewed elsewhere.

**Comment by yakky** [comment]
@czpython press that green button and enjoy the feeling

**Comment by czpython** [comment]
Here it is so we can all relive this moment over and over again:

1sppjqo - imgur

💯  👯 

Thanks to all involved.

**Pull request merged**